loudml/prometheus.py

```diff
@@ -0,0 +1,392 @@
+"""
+Prometheus module for Loud ML.
+
+ * PrometheusClient class handles connection to Prometheus
+ * PrometheusBucket class is capable to read and write timed data
+"""
+
+import logging
+import requests
+import json
+
+import numpy as np
+
+from voluptuous import (
+    Required,
+    Optional,
+    Boolean,
+)
+
+from . import (
+    errors,
+    schemas,
+)
+from loudml.misc import (
+    escape_quotes,
+    escape_doublequotes,
+    make_ts,
+    parse_addr,
+    str_to_ts,
+    ts_to_str,
+)
+from loudml.bucket import Bucket
+
+# Prometheus aggregation function:
+# https://prometheus.io/docs/prometheus/latest/querying/operators/#aggregation-operators
+# sum (calculate sum over dimensions)
+# min (select minimum over dimensions)
+# max (select maximum over dimensions)
+# avg (calculate the average over dimensions)
+# stddev (calculate population standard deviation over dimensions)
+# stdvar (calculate population standard variance over dimensions)
+# count (count number of elements in the vector)
+# count_values (count number of elements with the same value)
+# bottomk (smallest k elements by sample value)
+# topk (largest k elements by sample value)
+# quantile (calculate φ-quantile (0 ≤ φ ≤ 1) over dimensions)
+
+AGGREGATORS = {
+    'avg': 'avg',
+    'mean': 'avg',
+    'average': 'avg',
+    'stddev': 'stddev',
+    'std_dev': 'stddev',
+    'count': 'count',
+    'min': 'min',
+    'max': 'max',
+    'sum': 'sum',
+    'deriv': None,
+    'derivative': None,
+    'integral': None,
+    'med': None,
+    'median': None,
+    'mode': None,
+    '5percentile': None,
+    '10percentile': None,
+    '90percentile': None,
+    '95percentile': None
+}
+
+
+def _build_time_predicates(
+    from_date=None,
+    to_date=None,
+):
+    """
+    Build time range predicates
+    """
+    must = []
+
+    if from_date:
+        must.append("start={}".format(from_date))
+    if to_date:
+        must.append("end={}".format(to_date))
+
+    return "&".join(must)
+
+
+def _build_tags_predicates(match_all=None):
+    """
+    Build tags predicates
+    """
+    must = []
+
+    if match_all:
+        for condition in match_all:
+            must.append('{}="{}"'.format(condition['tag'], condition['value']))
+
+    return "{" + ", ".join(must) + "}"
```

Comment:
is the extra space intentional in ", " before join() ?

Reply:
it's typo, thank you